Added extensions and fixes for DBC features required by Proemion

johnbivens opened the pull request:
Fixes https://github.com/Proemion/quarks/issues/394, aside from integration into our build system (that bit is done in client code using this library, for instance applications).

johnbivens committed:
Repaired dbclint argument handling (double parse call).

johnbivens committed:
Reduce debug noise when parsing numbers.

johnbivens committed:
Use number in cm_bo and cm_sg grammars.

The use of integer was more correct, but integer is otherwise used
nowhere -- even for things like bit position or message length -- so
let's move to number to prevent the need for an additional integer
parser.

johnbivens committed:
Renamed CANsignal::receiver to CANsignal::receivers for consistency.

johnbivens committed:
Added missing calls to clear ecu_tokens after parsing a message or signal.

This caused the ecu_tokens vector to grow and grow under certain
circumstances, somtimes containing dozens of ECU tokens when parsing a
signal.

johnbivens committed:
Mux now uses an enum; corrected non-muxed signals in muxed messages.

It is possible for multiplexed messages to contain non-muxed signals,
which are always present regardless of the multiplexer value. These
signals were not handled correctly, because the fields CANsignal::mux
and CANsignal::muxNdx did not provide an indication of this. With the
new enum, this is now handled more gracefully.

The mux grammar was also renamed to muxer in an attempt to make the
separation of the mux types more obvious.

johnbivens committed:
Clear mux-related variables after parsing a signal.

In some cases, for instance when parsing a non-muxed signal within an
otherwise muxed message, not clearing these values caused improper mux
values to be stored for that signal.

johnbivens committed:
Improve user-friendliness of the CANsignal API with an endianness enum.

johnbivens committed:
ECUs named "Vector__XXX" should be ignored (magic name).

johnbivens committed:
CANsignal::muxNdx now uses boost::optional for better semantics.

johnbivens committed:
Added initial message and signal comment parsers (CM_ BO_, CM_ SG_).

johnbivens committed:
Snapshot: split ba_def_num into separate BO/SG/BU and INT/HED grammars.

This change was made to make parsing simpler for things like value
ranges for "GenMsgDelayTime" or "GenSigStartValue"; this may be
refactored in the future to reduce the need for multiple parsers.

johnbivens committed:
Support number and phrase for values in the ba_sg grammar.

johnbivens committed:
Added initial parsers for GenMsgCycleTime and GenSigStartValue values.

johnbivens committed:
Added some TODO comments.

johnbivens committed:
Minor code consistency cleanup.

johnbivens committed:
Store parsed message and signal comments.

johnbivens committed:
Minor cleanup: use quotes when printing parsed phrases.

johnbivens committed:
Removed unused field CANmessage::initValue.

johnbivens committed:
Store parsed global values for GenMsgCycleTime and GenSigStartValue in CANdb_t.

johnbivens committed:
Store parsed cycle times in CANmessage.

This change also renames the previously-unused field
CANmessage::updateCycle.

johnbivens committed:
Store parsed GenSigStartValue values in CANsignal using boost::any.

johnbivens committed:
Added support for VAL_ parsing and signal value descriptions.

johnbivens committed:
Added support for sig_val parsing and signal value types.

johnbivens committed:
Minor cleanup (comment).

johnbivens committed:
Support multiple transmitting ECUs in CANmessage and bo_tx_bu parsing.

johnbivens committed:
Simplified default initializers for boost::optionals in CANmessage/CANsignal.

johnbivens committed:
Added helpers to find CANdb_t message and signal iterators while parsing.

johnbivens committed:
Parse numbers as doubles by default and store doubles where appropriate.

double seems like a more reasonable choice, as the parser was previously
converting _all_ numbers to unsigned long long when parsing them. This
obviously broke floating point values found in the DBC. It seems DBC
files also use doubles when representing things like the maximum value
range for 64-bit (unsigned) integers, so double seems like a safer
choice overall.

Example from a DBC:

BO_ 3221225472 VECTOR__INDEPENDENT_SIG_MSG: 0 Vector__XXX
SG_ NeuesSignal_47 : 0|64@1- (1,0)
[-9.22337203685478E+018|9.22337203685478E+018] "" Vector__XXX
SG_ NeuesSignal_46 : 0|64@1+ (1,0) [0|1.84467440737096E+019] ""
Vector__XXX

johnbivens committed:
Use defined DBC enum values for signal value types.

This changes the existing CANsignalType enum class, but it seems that
was not used anywhere anyway.

johnbivens committed:
Minor formatting fix for consistency.

johnbivens committed:
Revert INT/HEX separation for BA_DEF_ BO_ and BA_DEF_ _SG_.

It seems the DBC format uses decimal numbers in both cases, regardless
of the INT/HEX specifier. In that case, there's currently no need to
differentiate between the two if the numbers are always interpreted in
the same way.

johnbivens committed:
Return to a combined ba_def_num rule definition and parsing.

This completely reverts the temporary split initially performed in
6311a5a and partially reverted in 2ada405. It again makes it possible to
parse lines of the form:

    BA_DEF_  "UseGMParameterIDs" INT 0 0;

Unfortunately, this also means the parser for ba_def_num needs to
determine whether or not the identifiers BO_, SG_, or BU_ are present in
the peg::SemanticValues parsed to the parsing lambda to differentiate
between the various forms.

johnbivens committed:
Fix out-of-bounds access to numbers when parsing VAL_.

johnbivens committed:
Extended the default CANmessage constructor with new optional fields.

johnbivens committed:
Corrected optional parsing of message ECUs.

johnbivens committed:
Minor code cleanup for consistency.

johnbivens committed:
Added a unit test for new/extended features.

saidinesh5 commented on src/cantypes.hpp:
```diff
@@ -32,20 +44,24 @@ struct CANsignal {
     CANmessage(std::uint32_t _id, const std::string& _name = "",
-        std::uint32_t _dlc = 0, const std::string& _ecu = "")
+        std::uint32_t _dlc = 0, const std::vector<std::string>& _ecus = { },
+        boost::optional<std::uint32_t> _cycleTime = boost::none,
```
is cycle time supposed to be 32bit or 64bit? In CanMessages object i used 63 bit unsigned ints for this.. but would be nice to use the same datatype everywhere.

johnbivens replied:
I will strongly assume 32-bit, because even 64-bit values are written as 32-bit numbers in the DBC file (as we saw when dealing with values). Anyway, this is milliseconds -- and I don't anticipate anyone using a cycle time of more than 4294967295ms = ~7.1 **weeks** :innocent:

saidinesh5 replied:
Updated my model and protobuf files for this.

saidinesh5 commented on src/cantypes.hpp:
```diff
@@ -5,23 +5,35 @@
+    CANsignalMuxType muxType{ CANsignalMuxType::NotMuxed };
+    // 16-bit for better debug printing via boost::optional
+    boost::optional<std::uint16_t> muxNdx{ boost::none };
+    boost::optional<boost::any> startValue { boost::none };
```
should this be boost:optional<double> ?

johnbivens replied:
Yeah, because the start value is optional. It is not part of a signal definition and only optionally appears later in the file, in a form like this:

```
BA_ "GenSigStartValue" SG_ 100 DRIVER_HEARTBEAT_cmd 3;
```

johnbivens committed:
Made necessary changes to support compilation using the C++11 standard.

johnbivens committed:
Resolved unusued variable compilation warning.

johnbivens committed:
Enable compile-time deactivation of logging via spdlog.

This makes it possible to use CANdb without the spdlog dependency,
simply by defining DISABLE_CDB_LOG. Log output via spdlog is obviously
lost when doing this.

saidinesh5 commented on src/cantypes.hpp:
```diff
@@ -32,20 +78,24 @@ struct CANsignal {
-    std::string ecu;
-    std::uint32_t updateCycle{ 0 };
-    std::string initValue{ "" };
+    std::vector<std::string> ecus;
```
This is transmitter right? Shouldn't it be just `std::string` ?
The CanSignal is what can have multiple receivers iirc: 

```message = BO_ message_id message_name ':' message_size transmitter {signal} ;```

```
signal = 'SG_' signal_name multiplexer_indicator ':' start_bit '|' signal_size '@' byte_order value_type '(' factor ',' offset ')' '[' minimum '|' maximum ']' unit receiver {',' receiver} ;
multiplexer_indicator = ' ' | 'M' | m multiplexer_switch_value ;
```

johnbivens replied:
You are correct that only a single transmitting ECU can be defined in the message definition itself -- but there is of course another optional line that can come later in the file to specify further transmitting ECUs for a message based on its CAN ID. In the following example, two transmitting ECUs are defined for the CAN ID with ID 500:

```
BO_TX_BU_ 500 : DRIVER,IO;
```

The original message definition only contained one transmitting ECU, as that's all the DBC format allows at the point of message definition:

```
BO_ 500 IO_DEBUG: 4 IO
```

saidinesh5 replied:
Updated my model for this

johnbivens committed:
Explicitly include the string header in dbc_grammar.hpp.in.

saidinesh5 commented on src/dbcparser.cpp:
```diff
@@ -73,6 +84,159 @@ std::string dos2unix(const std::string& data)
+inline std::vector<CANsignal>::iterator
+#endif
+    getSignalIteratorByMessageIdAndName(CANdb_t &canDb, uint32_t id,
+        const std::string& name, bool &found)
```
Just an api clean up suggestion,
here instead of `bool &found`, if it is `bool *found = nullptr`,
Then when reading the code, things will be more obvious when looking at the &, meaning this variable will be changed.
like:
`getSignalIteratorByMessageIdAndName(blah blah blah, &myVariable)`

saidinesh5 commented on src/dbcparser.cpp:
```diff
@@ -73,6 +84,159 @@ std::string dos2unix(const std::string& data)
+#if __cplusplus >= 201402L
+inline auto
+#else
+inline std::vector<CANsignal>::iterator
```
Why bother with if __cplusplus 14 at all here? We can directly use inline std::vector<CANsignal>::iterator right? I mean, if the auto keyword is not reducing the amount of typing we do, no point using auto right?

Also instead of bool found, if not found, we can simply return std::vector<CANsignal>::end() right?

johnbivens replied:
The original code was for C++14, so I made this change this way so as not to rewrite the upstream code (make the C++11 support optional). Not strictly necessary, but I was trying to be more upstream-friendly.

As for the return value, we don't always have an iterator to compare against -- so we can't always check if the return value is equal to the iterator end. That's the only reason I came up with this workaround `bool` solution in the first place.

johnbivens replied:
> The original code was for C++14, so I made this change this way so as not to rewrite the upstream code (make the C++11 support optional). Not strictly necessary, but I was trying to be more upstream-friendly.

Ehh... changing this after all just to be C++11-compliant.

johnbivens replied:
Changed with a force push after rebase :+1:

johnbivens committed:
Use pointers as success indicators when searching for message/signal iterators.

This is arguably a better API, making it more obvious to the calling
code that the pointer past will be changed.

saidinesh5 commented on src/dbcparser.cpp:
```diff
@@ -73,6 +84,159 @@ std::string dos2unix(const std::string& data)
+        cdb_debug("Found the message that needs the ECUs");
+        CANmessage updatedMessage(messageIt->first);
+        std::vector<CANsignal> existingSignals(messageIt->second);
+        while (!ecus.empty()) {
```
I think set insert/union operations here would make this code/api cleaner, so are we sure we need ecus to be a `std::vector<std::string>` ? or `std::unordered_set<std::string>` should also work?
Alternatively we can just use temporary `std::unordered_set<std::string>` here to do the merge operations.

johnbivens replied:
Maybe... but the upstream code works with `std::deque`, also to push and pop when parsing via cpp-peglib, so I really didn't want to fiddle with this too much.

saidinesh5 commented on src/dbcparser.cpp:
```diff
@@ -73,6 +84,159 @@ std::string dos2unix(const std::string& data)
+    }
+}
+
+void setSignalValueType(CANdb_t &canDb, uint32_t id,
```
return type bool maybe useful for these "setBlah" functions, especially when showing warnings

johnbivens replied:
As discussed on Slack, we'll revisit this later depending on how we want to do warning, error and exception handling (if what we already have isn't enough). But in general, you're right -- that would be a better API.

saidinesh5 commented on src/dbc_grammar.peg:
```diff
@@ -13,16 +13,16 @@ message                 <- 'BO_' s* number s* TOKEN ':' s number s TOKEN _ signa
 cm_bu                   <- < 'CM_' s* 'BU_' s* TOKEN s* phrase > ';' NewLine
-cm_bo                   <- < 'CM_' s* 'BO_' s* integer s* phrase > ';' NewLine
-cm_sg                   <- < 'CM_' s* 'SG_' s* integer s* TOKEN s* phrase > ';' NewLine*
+cm_bo                   <- < 'CM_' s* 'BO_' s* number s* phrase > ';' NewLine
```
Message id is supposed to be an unsigned integer.. so would changing it to number allow more values than what we accept?

``message = BO_ message_id message_name ':' message_size transmitter {signal} ;
message_id = unsigned_integer ; ```

johnbivens replied:
Yes, you're right. But this would require rewriting quite a bit more of this parser, which is why I just stuck with number -- see also the commit message in https://github.com/Proemion/CANdb/pull/1/commits/b3d97938620e82a63936e3241ff5bbf3c7ee98d4

saidinesh5 committed:
Make things build on MinGW

saidinesh5 committed:
Use C++11 instead of C++14

saidinesh5 reviewed (approved):
The code builds for windows and linux.
The API is mostly okay, with a few exceptions here and there that need to be cleaned up in a separate pull request. (Like reporting warnings and errors in a better way for eg.)
The parser mostly matches the spec at http://read.pudn.com/downloads766/ebook/3041455/DBC_File_Format_Documentation.pdf .
